test/Analysis/test-alias.mlir

```diff
@@ -216,10 +216,10 @@ tt.func @for_for_if(%lb : index, %ub : index, %step : index, %A : !tt.ptr<f16>,
   // CHECK-NEXT: %arg9 -> %cst_1
   // CHECK-NEXT: %0#0 -> %cst
   // CHECK-NEXT: %0#1 -> %cst_0
-  // CHECK-NEXT: %0#2 -> %cst_2,%cst_2
+  // CHECK-NEXT: %0#2 -> %cst_1,%cst_2,%cst_2
   %a_shared, %b_shared, %c_shared = scf.for %iv = %lb to %ub step %step iter_args(%a_shared = %a_shared_init, %b_shared = %b_shared_init, %c_shared = %c_shared_init) -> (tensor<128x32xf16, #A_SHARED>, tensor<128x32xf16, #A_SHARED>, tensor<128x32xf16, #A_SHARED>) {
     // CHECK-NEXT: %arg11 -> %cst_1,%cst_2,%cst_2
-    // CHECK-NEXT: %1 -> %cst_2,%cst_2
+    // CHECK-NEXT: %1 -> %cst_1,%cst_2,%cst_2
```

Comment:
@Jokeren if you wanna double check, but these makes sense to me!

Reply:
Checking... %1 should be associated with %cst_1 even without the commit.

Reply:
hmm, I don't think %1 should be associated with %cst_1

```
module {
  tt.func @for_for_if(%arg0: index, %arg1: index, %arg2: index, %arg3: !tt.ptr<f16, 1>, %arg4: !tt.ptr<f16, 1>, %arg5: i1) {
    %cst = arith.constant dense<0.000000e+00> : tensor<128x32xf16, #shared>
    %cst_0 = arith.constant dense<0.000000e+00> : tensor<128x32xf16, #shared>
    %cst_1 = arith.constant dense<0.000000e+00> : tensor<128x32xf16, #shared>
    %0:3 = scf.for %arg6 = %arg0 to %arg1 step %arg2 iter_args(%arg7 = %cst, %arg8 = %cst_0, %arg9 = %cst_1) -> (tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>) {
      %1 = scf.for %arg10 = %arg0 to %arg1 step %arg2 iter_args(%arg11 = %arg9) -> (tensor<128x32xf16, #shared>) {
        %2 = scf.if %arg5 -> (tensor<128x32xf16, #shared>) {
          %cst_2 = arith.constant dense<0.000000e+00> : tensor<128x32xf16, #shared>
          scf.yield %cst_2 : tensor<128x32xf16, #shared>
        } else {
          %cst_2 = arith.constant dense<0.000000e+00> : tensor<128x32xf16, #shared>
          scf.yield %cst_2 : tensor<128x32xf16, #shared>
        }
        scf.yield %2 : tensor<128x32xf16, #shared>
      }
      scf.yield %arg7, %arg8, %1 : tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>
    }
    tt.return
  }
}
```

What's your output IR?

Reply:
```
    %0:3 = scf.for %arg6 = %arg0 to %arg1 step %arg2 iter_args(%arg7 = %cst, %arg8 = %cst_0, %arg9 = %cst_1) -> (tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>, tensor<128x32xf16, #shared>) {
      %1 = scf.for %arg10 = %arg0 to %arg1 step %arg2 iter_args(%arg11 = %arg9) -> (tensor<128x32xf16, #shared>) {
        %2 = scf.if %arg5 -> (tensor<128x32xf16, #shared>) {
```

If the outer loop is executed but the not the inner loop, then `%1 == %cst_1`.

Do you expect the algorithm here to realize that the two loops have the same bounds and so this is impossible?

Reply:
Thanks. I get that, maybe it's due to the update in LLVM's dataflow analyzer? Our analysis module hasn't done anything special.

Reply:
We get iteration arguments but not initialization arguments, so it assumed that the loop must run at least once I think

Reply:
I do remembered there has been some changes to scf.for's argument but haven't paid close attention

Reply:
There was never a guarantee that the loop would run at least once, I suspect there was a fix upstream in the regionbranchinterface for that.

Reply:
I wouldn't be surprised if it was somehow related to the work in this area: https://reviews.llvm.org/D157506

Reply:
Maybe: https://reviews.llvm.org/D156376 "[MLIR][analysis] Fix error in the sparse backward dataflow analysis"

Or better:  https://reviews.llvm.org/D155822

```
  [mlir] allow region branch spec from parent op to itself
    
    RegionBranchOpInterface did not allow the operation with regions to
    specify itself as successors. Therefore, this implied that the control
    is always transferred to a region before being transferred back to the
    parent op. Since the region can only transfer the control back to the
    parent op from a terminator, this transitively implied that the first
    block of any region with a RegionBranchOpInterface is always executed
    until the terminator can transfer the control flow back. This is
    trivially false for any conditional-like operation that may or may not
    execute the region, as well as for loop-like operations that may not
    execute the body.
```

Reply:
Makes sense to me